flang/lib/Optimizer/Transforms/ArrayValueCopy.cpp

```diff
@@ -541,18 +540,58 @@ class ArrayUpdateConversion : public mlir::OpRewritePattern<ArrayUpdateOp> {
     } else {
       LLVM_DEBUG(llvm::outs() << "No, conflict wasn't found\n");
       rewriter.setInsertionPoint(op);
+      auto coorTy = getEleTy(load.getType());
       auto coor = rewriter.create<ArrayCoorOp>(
-          loc, getEleTy(load.getType()), load.memref(), load.shape(),
-          load.slice(),
+          loc, coorTy, load.memref(), load.shape(), load.slice(),
           originateIndices(loc, rewriter, load.shape(), update.indices()),
-          load.lenParams());
-      rewriter.create<fir::StoreOp>(loc, update.merge(), coor);
+          load.typeparams());
+      auto input = update.merge();
+      if (auto inEleTy = fir::dyn_cast_ptrEleTy(input.getType())) {
+        auto outEleTy = fir::unwrapSequenceType(update.getType());
+        if (auto inChrTy = inEleTy.dyn_cast<fir::CharacterType>()) {
+          assert(outEleTy.isa<fir::CharacterType>());
+          fir::factory::genCharacterCopy(input, recoverCharLen(input), coor,
+                                         recoverCharLen(coor), rewriter, loc);
+        } else if (inEleTy.isa<fir::RecordType>()) {
+          TODO(loc, "copy derived type");
+        } else {
+          llvm::report_fatal_error("not a legal reference type");
+        }
+      } else {
+        rewriter.create<fir::StoreOp>(loc, input, coor);
+      }
     }
     update.replaceAllUsesWith(load.getResult());
     rewriter.replaceOp(update, load.getResult());
     return mlir::success();
   }
 
+  static llvm::SmallVector<mlir::Value> recoverTypeParams(mlir::Value val) {
```

Comment:
You might have to handle the fir.box  case here. `ArrayLoadOp` and `ArrayCoorOp` can take fir.box argument, and I am not sure what should be the rule about typeparams in this case (I am not sure it occured yet). For fir.box memref, we decided not to have extents arguments on the basis that it was duplicating the ones from the fir.box argument. One could argue the same for the length parameters.

Reply:
Yes. We do have to handle boxes.